File: day55/isValidSudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        row = [{} for _ in range(9)]
        col = [{} for _ in range(9)]
        box = [{} for _ in range(9)]

        for i in range(9):
            for j in range(9):
                if board[i][j] == ".": continue
                
                num = board[i][j]
                k = i // 3 * 3 + j // 3
                row[i][num] = row[i].get(num, 0) + 1
                col[j][num] = col[j].get(num, 0) + 1
                box[k][num] = box[k].get(num, 0) + 1

                if row[i][num] > 1 or col[j][num] > 1 or box[k][num] > 1: return False
                
        return True

# class Solution:
#     def isValidSudoku(self, board: List[List[str]]) -> bool:
#         row = [set() for _ in range(9)]
#         col = [set() for _ in range(9)]
#         box = [[set() for i in range(3)] for j in range(3)]

#         for i in range(9):
#             for j in range(9):
#                 if board[i][j] == ".": continue
                
#                 cur = board[i][j]
#                 if cur not in row[i]: row[i].add(cur)
#                 else: return False

#                 if cur not in col[j]: col[j].add(cur)
#                 else: return False

#                 if cur not in box[i // 3][j // 3]: box[i // 3][j // 3].add(cur)
#                 else: return False

#         return True

File: day55/isValidSudoku.java
/*
36. 有效的数独
判断一个 9x9 的数独是否有效。只需要根据以下规则，验证已经填入的数字是否有效即可。
数字 1-9 在每一行只能出现一次。
数字 1-9 在每一列只能出现一次。
数字 1-9 在每一个以粗实线分隔的 3x3 宫内只能出现一次。


来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/valid-sudoku
*/

class Solution {
    public boolean isValidSudoku(char[][] board) {
        Map<Integer, Integer>[] row = new HashMap[9];
        Map<Integer, Integer>[] col = new HashMap[9];
        Map<Integer, Integer>[] box = new HashMap[9];
        
        for (int i = 0; i < 9; i++) {
            row[i] = new HashMap<Integer, Integer>();
            col[i] = new HashMap<Integer, Integer>();
            box[i] = new HashMap<Integer, Integer>();

        }

        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.') continue;

                int num = (int) board[i][j]; 
                int k = i / 3 * 3 + j / 3;

                row[i].put(num, row[i].getOrDefault(num, 0) + 1);
                col[j].put(num, col[j].getOrDefault(num, 0) + 1);
                box[k].put(num, box[k].getOrDefault(num, 0) + 1);

                if (row[i].get(num) > 1 || col[j].get(num) > 1 || box[k].get(num) > 1)
                    return false;

            }
        }

        return true;
    }
}
